pandas/tseries/frequencies.py

```diff
@@ -174,14 +174,12 @@ def infer_freq(
         return inferer.get_freq()
 
     if isinstance(index, Index) and not isinstance(index, DatetimeIndex):
+        values = index._values
         if index.dtype.kind in "iufcb":
             raise TypeError(
                 f"cannot infer freq from a non-convertible index of dtype {index.dtype}"
             )
-        # error: Incompatible types in assignment (expression has type
-        # "Union[ExtensionArray, ndarray[Any, Any]]", variable has type
-        # "Union[DatetimeIndex, TimedeltaIndex, Series, DatetimeLikeArrayMixin]")
-        index = index._values  # type: ignore[assignment]
+        index = values
```

Comment:
I'm surprised this works but I feel this might potentially hide bugs/inconsistencies in the type annotations. I think this mypy warning is "correct" given the type annotations.

I think this if statement could be `if isinstance(index, TimedeltaIndex):` If that is the case, `TimedeltaIndex._values` should be of type `TimedeltaArray` which is a sub-class of `DatetimeLikeArrayMixin` - meaning it wouldn't trigger this error. @jbrockmendel 

If the above is the case, I think there are two ways to fix it:

- simplify the if condition to `if isinstance(index, TimedeltaIndex):` and use `index = cast(TimedeltaArray, index._values)`
- or better, annotate `TimedeltaIndex._values` as `TimedeltaArray` (and still change the if condition)

Reply:
> or better, annotate TimedeltaIndex._values as TimedeltaArray

seems reasonable

> I think this if statement could be if isinstance(index, TimedeltaIndex)

No, TDI is specifically handled a few lines up in the `elif is_timedelta64_dtype(index.dtype):` section.  Looks like this bit is for object dtype.  though id be OK with deprecating allowing that here